### PR TITLE
Custom window frame support extended to built-in window functions and some refactors

### PR DESCRIPTION
**Pull request by mustafasrepo**
# Which issue does this PR close?

Closes #.

 # Rationale for this change

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/physical-expr/src/window/aggregate.rs [hunk 1]
```diff
@@ -109,17 +88,85 @@ impl WindowExpr for AggregateWindowExpr {
...
         // Sort values, this will make the same partitions consecutive. Also, within the partition
         // range, values will be sorted.
+        let order_bys =
+            &order_columns[self.partition_by.len()..order_columns.len()].to_vec();
+        let window_frame = match (&order_bys[..], &self.window_frame) {
+            ([column, ..], None) => {
+                // OVER (ORDER BY a) case
+                // We create an implicit window for ORDER BY.
+                let empty_bound = ScalarValue::try_from(column.data_type())?;
+                Some(Arc::new(WindowFrame {
```
Comment (ozankabak) [inline datafusion/physical-expr/src/window/aggregate.rs:101]: Would `WindowFrame::default()` work here?

File: datafusion/physical-expr/src/window/aggregate.rs [hunk 2]
```diff
@@ -109,17 +88,85 @@ impl WindowExpr for AggregateWindowExpr {
...
+        };
         let results = partition_points
             .iter()
             .map(|partition_range| {
-                let mut window_accumulators = self.create_accumulator()?;
-                Ok(vec![window_accumulators.scan(
-                    &values,
-                    &array_refs,
-                    partition_range,
+                // let mut accumulator = self.aggregate.create_accumulator()?;
```
Comment (ozankabak) [inline datafusion/physical-expr/src/window/aggregate.rs:112]: Possible leftover from debugging, remove.

File: datafusion/physical-expr/src/window/aggregate.rs [hunk 3]
```diff
@@ -109,17 +88,85 @@ impl WindowExpr for AggregateWindowExpr {
...
+
+                for i in 0..length {
+                    let cur_range = self.calculate_range(
+                        &window_frame,
+                        &slice_order_bys,
+                        &sort_options,
+                        length,
+                        i,
+                    )?;
+                    // println!("cur range: {:?}", cur_range);
```
Comment (ozankabak) [inline datafusion/physical-expr/src/window/aggregate.rs:139]: Possible leftover from debugging, remove.

File: datafusion/physical-expr/src/window/built_in.rs [hunk 4]
```diff
@@ -80,16 +86,78 @@ impl WindowExpr for BuiltInWindowExpr {
...
+                    &order_columns[self.partition_by.len()..order_columns.len()].to_vec();
+
+                let mut ranges = vec![];
+
+                let window_frame = match (&order_bys[..], &self.window_frame) {
+                    ([column, ..], None) => {
+                        // OVER (ORDER BY a) case
+                        // We create an implicit window for ORDER BY.
+                        let empty_bound = ScalarValue::try_from(column.data_type())?;
+                        Some(Arc::new(WindowFrame {
```
Comment (ozankabak) [inline datafusion/physical-expr/src/window/built_in.rs:112]: Would `WindowFrame::default()` work here?

File: datafusion/physical-expr/src/window/built_in.rs [hunk 5]
```diff
@@ -80,16 +86,78 @@ impl WindowExpr for BuiltInWindowExpr {
...
+            (_, true) => {
+                let sort_options: Vec<SortOptions> =
+                    self.order_by.iter().map(|o| o.options).collect();
+                let columns = self.sort_columns(batch)?;
+                let order_columns: Vec<&ArrayRef> =
+                    columns.iter().map(|s| &s.values).collect();
+                // Sort values, this will make the same partitions consecutive. Also, within the partition
+                // range, values will be sorted.
+                let order_bys =
+                    &order_columns[self.partition_by.len()..order_columns.len()].to_vec();
```
Comment (ozankabak) [inline datafusion/physical-expr/src/window/built_in.rs:103]: Can we just use:
```rust
let order_bys = &order_columns[self.partition_by.len()..].to_vec();
```

File: datafusion/physical-expr/src/window/aggregate.rs [hunk 6]
```diff
@@ -109,17 +88,85 @@ impl WindowExpr for AggregateWindowExpr {
         let values = self.evaluate_args(batch)?;
 
         let columns = self.sort_columns(batch)?;
-        let array_refs: Vec<&ArrayRef> = columns.iter().map(|s| &s.values).collect();
+        let order_columns: Vec<&ArrayRef> = columns.iter().map(|s| &s.values).collect();
         // Sort values, this will make the same partitions consecutive. Also, within the partition
         // range, values will be sorted.
+        let order_bys =
+            &order_columns[self.partition_by.len()..order_columns.len()].to_vec();
```
Comment (ozankabak) [inline datafusion/physical-expr/src/window/aggregate.rs:95]: Can we just use:
```rust
let order_bys = &order_columns[self.partition_by.len()..].to_vec();
```

File: datafusion/physical-expr/src/window/built_in.rs [hunk 7]
```diff
@@ -80,16 +86,78 @@ impl WindowExpr for BuiltInWindowExpr {
...
+                        ranges.push(Range {
+                            start: res.0,
+                            end: res.1,
+                        });
+                    }
+                }
+
+                ranges
+                    .iter()
+                    .map(|elem| evaluator.evaluate_inside_range(elem.clone()).unwrap())
```
Comment (ozankabak) [inline datafusion/physical-expr/src/window/built_in.rs:149]: Do we have a guarantee that the `unwrap` call will not fail?

File: datafusion/physical-expr/src/window/nth_value.rs [hunk 8]
```diff
@@ -142,53 +140,34 @@ pub(crate) struct NthValueEvaluator {
...
-                    })
-                    .collect::<Result<Vec<_>>>()?
-                    .into_iter()
-                    .flatten();
-                ScalarValue::iter_to_array(values)
+                let value = ScalarValue::try_from_array(arr, range.end - 1)?;
+                Ok(value.to_array_of_size(1))
             }
             NthValueKind::Nth(n) => {
                 let index = (n as usize) - 1;
```
Comment (ozankabak) [inline datafusion/physical-expr/src/window/nth_value.rs:164]: Is `n = 0` a possible case here? If so, does the code still work?

**Review by ozankabak — CHANGES_REQUESTED** [review]
Let's go through my initial reviews, I will review again after.

File: datafusion/physical-expr/src/window/nth_value.rs (hunk 8 above)
Comment (mustafasrepo) [reply datafusion/physical-expr/src/window/nth_value.rs:164]: Postgre expects nth_value argument to be larger than 0 according to [this link](https://www.postgresql.org/docs/current/functions-window.html). I have tried the nth_value(0) query on dbfiddle it gave  

> Query Error: error: argument of nth_value must be greater than zero

In datafusion nth_value(0) case raises [an error](https://github.com/apache/arrow-datafusion/blob/71f05a3175e931943276a8778162e37504e00503/datafusion/physical-expr/src/window/nth_value.rs#L89) also
I have added a comment also saying `// We are certain that n > 0`

File: datafusion/physical-expr/src/window/built_in.rs (hunk 7 above)
Comment (mustafasrepo) [reply datafusion/physical-expr/src/window/built_in.rs]: I have removed unwrap

**Review by ozankabak — APPROVED** [review]

File: datafusion/physical-expr/src/window/built_in.rs [hunk 9]
```diff
@@ -80,11 +86,59 @@ impl WindowExpr for BuiltInWindowExpr {
...
+            } else {
+                self.window_frame.clone()
+            };
+            for partition_range in &partition_points {
+                let length = partition_range.end - partition_range.start;
+                let slice_order_bys = order_bys
+                    .iter()
+                    .map(|v| v.slice(partition_range.start, length))
+                    .collect::<Vec<_>>();
+                // We iterate on each row to calculate window frame boundaries
```
Comment (ozankabak) [inline datafusion/physical-expr/src/window/built_in.rs:117]: Is this comment accurate?

**Review by ozankabak — APPROVED** [review]
LGTM. I sent a few simplifying changes too. I don't fully understand one comment (for which I left a review), apart from that -- looks good.

About that specific comment: Isn't iterating on all rows just to calculate frame boundaries very inefficient? Is this really what we do (and what you mean)? Let's clarify this, and if there is anything to fix/improve, fix it. Otherwise, we can just make the comment clearer, compose a PR body, and create an upstream PR to get community review.

File: datafusion/physical-expr/src/window/partition_evaluator.rs [hunk 10]
```diff
@@ -83,4 +87,11 @@ pub trait PartitionEvaluator {
             "evaluate_partition_with_rank is not implemented by default".into(),
         ))
     }
+
+    /// evaluate the partition evaluator against the partition but with rank
```
Comment (ozankabak) [inline datafusion/physical-expr/src/window/partition_evaluator.rs:91]: This comment is the same with the one above. This function needs its own comment.

Comment (mustafasrepo) [reply datafusion/physical-expr/src/window/partition_evaluator.rs]: Changed comment

File: datafusion/physical-expr/src/window/built_in.rs (hunk 9 above)
Comment (mustafasrepo) [reply datafusion/physical-expr/src/window/built_in.rs]: Changed the comment to make it more clear

**Comment by ozankabak** [comment]
Closing as this merged upstream.

**Pull request closed without merging**